### ISSUE TITLE
Check PHP 5.5's password_hash() function

### ISSUE THREAD

**Issue by ozh** [YOURLS/YOURLS#1434]
TODO: compare @nicwaller's PR with phpass with future function password_hash() and its compat lib maintained here: https://github.com/ircmaxell/password_compat

Edit: this function is PHP >= 5.3.7. Ponder when stats from api.yourls.org indicate it's OK to deprecated 5.2

**Comment by My1** [comment]
I totally agree. PHPass itself even recommends this and ALL PHP versions that are supported in any way have this thing.

**Comment by My1** [comment]
nzw considering 5.2 has been dropped with the 1.7.2 wouldnt it be great to implement it for the next release?

**Comment by My1** [comment]
by the way I am hacking something together to get this working, I know, don't hack core (there is the `is_valid_user` filter for overriding the auth, I know), but how am I supposed to have it

1) skip the hashing of password_hash hashes while still letting plain passwords through (change behavior of `yourls_has_cleartext_passwords`)?
2) change the hashing method of `yourls_hash_passwords_now` (not actually done yet, but will be needed)

at least as far as I can see there arent really any filters to use there.

**Comment by ozh** [comment]
@My1 if you're aiming at coding a plugin and you find yourself missing some filters, feel free to submit a PR to add them.

This said, this issue opened *more than 5 years ago* was about to change core itself, so, in this case, hacking core *is* what's expected.

**Comment by My1** [comment]
well since I already have a mostly working core hack (I can login and I made an albeit dirty [yes it will get some cleanup] way to ignore the passwords, I could also try a core PR as soon as I am finished.

also as soon as PHP 5.3.7 (or rather any version with the `$2y` fix will be set as minimum (instead of the current 5.3) I could even try to grill phpass and switch over to password_compat, and if 5.5 would be minimum we could even go one step further and have no library actually needed for this, but this will probably take a while)

https://github.com/ircmaxell/password_compat

the best thing is that it doesnt need any monster setup just include the one file and you're ready (or if you like composer, use that)

**Comment by My1** [comment]
well that makes a LOT of things easier, I could add 2 little values for the hash and the config and the rest will be just done with password_hash(), old PHPpass hashes can be converted back to the original form. (although it might not bad to keep phpass so users dont need to reset their passwords or what do you think)

(btw why should the $ signs even be an issue, considering you use `'` in the config file the $ signs in the hash shouldnt be interpreted by PHP as anything)

-> https://github.com/YOURLS/YOURLS/blob/master/includes/functions-auth.php#L170

**Comment by My1** [comment]
Progress Update. got the whole "rehash if the algo settings are different" fun done, that also should nicely convert old hashes like PHPass and md5. (also 2 new constants for the config regarding the algo and options.

~~~php

define ('YOURLS_PASSWORD_ALGO',constant('PASSWORD_ARGON2ID'));

//set password options depending on the algo
if(constant('YOURLS_PASSWORD_ALGO') === constant('PASSWORD_BCRYPT')) {
  define('YOURLS_PASSWORD_OPTIONS',['cost' => 10]);
}
elseif(constant('YOURLS_PASSWORD_ALGO') === constant('PASSWORD_ARGON2I') || constant('YOURLS_PASSWORD_ALGO') === constant('PASSWORD_ARGON2ID')) {
  define('YOURLS_PASSWORD_OPTIONS',['threads' => 4,'memory_cost' => 1<<18,'time_cost' => 1]);
}
else {
  define('YOURLS_PASSWORD_OPTIONS',[]);
}

~~~

(yes I fully know that die is not an option, these are just in for testing although can you tell me how to have yourls output error messages to the user

~~~php

function yourls_check_password_hash( $user, $submitted_password ) {
	global $yourls_user_passwords;
	
	if( !isset( $yourls_user_passwords[ $user ] ) ) {
    // dont't let users without password in
		return false;
  }
  $password_check=false;
  if(preg_match('#^(?>\$[A-Za-z0-9,/=.+]+){3,5}$#',$yourls_user_passwords[ $user ])) { // use # as regex boundary to not need escaping, check for modular crypt formar
    $password_check = password_verify($submitted_password,$yourls_user_passwords[ $user ]);
    if(!$password_check) {
      if(password_get_info( $yourls_user_passwords[ $user ] ) [ "algoName" ] == "unknown") {
        //post some error about an unknown password algo
      }
    }
  }
	elseif ( yourls_has_phpass_password( $user ) ) {
		// Stored password is hashed with phpass
		list( , $hash ) = explode( ':', $yourls_user_passwords[ $user ] );
		$hash = str_replace( '!', '$', $hash );
		$password_check = yourls_phpass_check( $submitted_password, $hash );
	} else if( yourls_has_md5_password( $user ) ) {
		// Stored password is a salted md5 hash: "md5:<$r = rand(10000,99999)>:<md5($r.'thepassword')>"
		list( , $salt, ) = explode( ':', $yourls_user_passwords[ $user ] );
		$password_check = $yourls_user_passwords[ $user ] == 'md5:'.$salt.':'.md5( $salt . $submitted_password );
	}
  else {
		// Password stored in clear text
		$password_check = $yourls_user_passwords[ $user ] == $submitted_password;
	}
  //only go to rehash if we actually had a correct password
  if ($password_check) {
  
    $unfiltered_needs_rehash=password_needs_rehash($yourls_user_passwords[ $user ],constant('YOURLS_PASSWORD_ALGO'),constant ('YOURLS_PASSWORD_OPTIONS'));
    
    //plugins can come in, get original user/pw via $_REQUEST, rehash and set to false
    $needs_rehash=yourls_apply_filter('password_needs_rehash',$unfiltered_needs_rehash);
    
    if($needs_rehash) {
      $config_file=constant('YOURLS_CONFIGFILE');
      if( !is_readable( $config_file ) ) {
        //post some error
        echo $config_file.PHP_EOL;
        die('conf not readable');
      }
      if( !is_writable( $config_file ) ) {
        //post some error
        die('conf not writable');
      }
      $configdata = file_get_contents( $config_file );
      if( $configdata == false ) {
        die('could not get conf');
      }
      $quotes = "'" . '"';
      $hash = password_hash($submitted_password, constant ('YOURLS_PASSWORD_ALGO'), constant('YOURLS_PASSWORD_OPTIONS'));
      $pattern = "/[$quotes]${user}[$quotes]\s*=>\s*[$quotes]" . preg_quote( $yourls_user_passwords[ $user ], '/' ) . "[$quotes]/";
      //make sure preg_replace doesnt do anything ($ and number means replace it with something that was found)
      $preg_hash=str_replace('$','\$',$hash);
      $replace = "'$user' => '$preg_hash'";
      $count = 0;
      
      $configdata = preg_replace( $pattern, $replace, $configdata, -1, $count );
      
      // There should be exactly one replacement. Otherwise, fast fail.
      if ( $count != 1 ) {
        yourls_debug_log( "Problem with preg_replace for password hash of user $user" );
        die('preg_replace problem');
      }
      $success = file_put_contents( $config_file, $configdata );
      if ( $success === FALSE ) {
        yourls_debug_log( 'Failed writing to ' . $config_file );
        die('could not write conf');
      }
    }
  }
  
  return $password_check;
}

~~~

is that thing okay or does it have problems?

**Comment by My1** [comment]
> why you think your approach is better than original code?

what do you mean by "original code" is my code fake in some way? I basically went ahead and used what's already available in yourls and built upon that (for example the whole replace the password in the config file part is mostly copied from `hash_passwords_now()` while accounting for the fact we are only doing 1 user).

regarding unit tests, I honestly don't know much about those, when I wanna test stuff I insert debug echos or logs and test and later remove those

**Comment by dgw** [comment]
> As of now (1.7.4+) YOURLS min req is PHP 5.6

Perhaps I should take this to a separate issue, but I think we should bump to 1.8. Changing a minimum dependency version could force a system upgrade, depending on the user's environment and how old/out-of-date it is, so this fits best in a non-patch release.

**Comment by My1** [comment]
definitely. makes a lot of sense.

**Comment by ozh** [comment]
> what do you mean by "original code" is my code fake in some way?

? Errr... No.... I mean the original code you're willing to replace/update.....

**Comment by My1** [comment]
oh, the missing "the" before "original code" in the part 

> why you think your approach is better than original code

kinda tripped me I guess. it sounded like "original code" was something special or what, not sure.

1) well I would say that the main part of the check password code is not "better" but it just has password_hash in as an option allowing for secure and most importantly standard hashes supported by PHP itself without any libraries or weird tricks, also in newer versions of PHP (5.3.7 onwards) PHPass drops in bcrypt anyway, also if the plan will be (in the future) to drop the older hash methods (5 digit salted single md5 is not really the safest bet out there) the whole password overview is (relatively) easy because PHP takes care of that, and it allows us to drop libraries, and I would guess less dependencies at no or marginal cost is a good thing.

2) it allows for changing the algo and cost (depending what PHP offers in any given version)

3) the code also allows for updates of the password hash, for example if one wants to raise the cost to counteract higher computing power of attackers with relative ease as it can rehash during login if needed or change the algorithm because it has become weak and all without much effort.

4) the rehash part has a filter inside so plugins can also go and use their own password hash algos and use the filter to update the password hash.

**Comment by ozh** [comment]
Closed via #3232
